fl4health/servers/base_server.py

```diff
@@ -44,30 +50,89 @@ def __init__(
+                provided model. Defaults to None.
+            intermediate_server_state_dir (Path): A directory to store and load state from for the server
+                during an FL experiment. This allows for the saving of server state in case federated training is
+                interrupted and needs to be restarted from the same point. If none, then no state is saved during each
+                server round. Defaults to None.
+            on_init_parameters_config_fn (Optional[Callable[[int], Dict[str, Scalar]]], optional):
+                Function used to configure how one asks a client to provide parameters from which to initialize all
+                other clients by providing a Config dictionary. If this is none, then a blank config is sent with the
+                parameter request (which is default behavior for flower servers). Defaults to None.
+            server_name (Optional[str], optional): An optional string name to uniquely identify server.
+                Defaults to None.
             accept_failures (bool, optional): Determines whether the server should accept failures during training or
                 evaluation from clients or not. If set to False, this will cause the server to shutdown all clients
                 and throw an exception. Defaults to True.
         """
 
         super().__init__(client_manager=client_manager, strategy=strategy)
+        self.fl_config = fl_config
+        self.server_model = model
+        self.on_init_parameters_config_fn = on_init_parameters_config_fn
         self.checkpointer = [checkpointer] if isinstance(checkpointer, TorchCheckpointer) else checkpointer
+        # To facilitate model rehydration from server-side state for checkpointing
+        self.parameter_exchanger = parameter_exchanger
         self.server_name = server_name if server_name is not None else generate_hash()
         self.accept_failures = accept_failures
 
+        self.per_round_checkpointer: PerRoundCheckpointer | None
+
+        if intermediate_server_state_dir is not None:
+            log(
+                WARNING,
+                "intermediate_server_state_dir is not None. Creating PerRoundCheckpointer. This functionality still "
+                "experimental and only supported for BasicClient and NnunetClient currently.",
+            )
+            self.per_round_checkpointer = PerRoundCheckpointer(
+                intermediate_server_state_dir, Path(f"{self.server_name}.ckpt")
+            )
+        else:
+            self.per_round_checkpointer = None
+
+        self.current_round: int
+        self.history: History
+
         # Initialize reporters with server name information.
         self.reports_manager = ReportsManager(reporters)
         self.reports_manager.initialize(id=self.server_name)
+        self._log_fl_config()
+
+    def update_before_fit_round(self, server_round: int, timeout: Optional[float] = None) -> None:
```

Comment:
This function takes the place of the `initialize` functionality that was present in the FlServerWithIntiailizer class. It's called before every fit round starts. In the case of nnunet, it simply does the initialization process once (when server_round == 1) and then does nothing there after.